File: src/main/java/com/merge/api/resources/hris/types/PayrollRun.java
package com.merge.api.resources.hris.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = PayrollRun.Builder.class)
public final class PayrollRun {
    private final Optional<String> id;

    private final Optional<String> remoteId;

    private final Optional<PayrollRunRunState> runState;

    private final Optional<PayrollRunRunType> runType;

    private final Optional<OffsetDateTime> startDate;

    private final Optional<OffsetDateTime> endDate;

    private final Optional<OffsetDateTime> checkDate;

    private final Optional<Boolean> remoteWasDeleted;

    private final Optional<OffsetDateTime> createdAt;

    private final Optional<OffsetDateTime> modifiedAt;

    private final Optional<Map<String, JsonNode>> fieldMappings;

    private final Optional<List<RemoteData>> remoteData;

    private PayrollRun(
            Optional<String> id,
            Optional<String> remoteId,
            Optional<PayrollRunRunState> runState,
            Optional<PayrollRunRunType> runType,
            Optional<OffsetDateTime> startDate,
            Optional<OffsetDateTime> endDate,
            Optional<OffsetDateTime> checkDate,
            Optional<Boolean> remoteWasDeleted,
            Optional<OffsetDateTime> createdAt,
            Optional<OffsetDateTime> modifiedAt,
            Optional<Map<String, JsonNode>> fieldMappings,
            Optional<List<RemoteData>> remoteData) {
        this.id = id;
        this.remoteId = remoteId;
        this.runState = runState;
        this.runType = runType;
        this.startDate = startDate;
        this.endDate = endDate;
        this.checkDate = checkDate;
        this.remoteWasDeleted = remoteWasDeleted;
        this.createdAt = createdAt;
        this.modifiedAt = modifiedAt;
        this.fieldMappings = fieldMappings;
        this.remoteData = remoteData;
    }

    @JsonProperty("id")
    public Optional<String> getId() {
        return id;
    }

    /**
     * @return The third-party API ID of the matching object.
     */
    @JsonProperty("remote_id")
    public Optional<String> getRemoteId() {
        return remoteId;
    }

    /**
     * @return The state of the payroll run
     * <ul>
     * <li><code>PAID</code> - PAID</li>
     * <li><code>DRAFT</code> - DRAFT</li>
     * <li><code>APPROVED</code> - APPROVED</li>
     * <li><code>FAILED</code> - FAILED</li>
     * <li><code>CLOSED</code> - CLOSED</li>
     * </ul>
     */
    @JsonProperty("run_state")
    public Optional<PayrollRunRunState> getRunState() {
        return runState;
    }

    /**
     * @return The type of the payroll run
     * <ul>
     * <li><code>REGULAR</code> - REGULAR</li>
     * <li><code>OFF_CYCLE</code> - OFF_CYCLE</li>
     * <li><code>CORRECTION</code> - CORRECTION</li>
     * <li><code>TERMINATION</code> - TERMINATION</li>
     * <li><code>SIGN_ON_BONUS</code> - SIGN_ON_BONUS</li>
     * </ul>
     */
    @JsonProperty("run_type")
    public Optional<PayrollRunRunType> getRunType() {
        return runType;
    }

    /**
     * @return The day and time the payroll run started.
     */
    @JsonProperty("start_date")
    public Optional<OffsetDateTime> getStartDate() {
        return startDate;
    }

    /**
     * @return The day and time the payroll run ended.
     */
    @JsonProperty("end_date")
    public Optional<OffsetDateTime> getEndDate() {
        return endDate;
    }

    /**
     * @return The day and time the payroll run was checked.
     */
    @JsonProperty("check_date")
    public Optional<OffsetDateTime> getCheckDate() {
        return checkDate;
    }

    /**
     * @return Indicates whether or not this object has been deleted in the third party platform.
     */
    @JsonProperty("remote_was_deleted")
    public Optional<Boolean> getRemoteWasDeleted() {
        return remoteWasDeleted;
    }

    @JsonProperty("created_at")
    public Optional<OffsetDateTime> getCreatedAt() {
        return createdAt;
    }

    /**
     * @return This is the datetime that this object was last updated by Merge
     */
    @JsonProperty("modified_at")
    public Optional<OffsetDateTime> getModifiedAt() {
        return modifiedAt;
    }

    @JsonProperty("field_mappings")
    public Optional<Map<String, JsonNode>> getFieldMappings() {
        return fieldMappings;
    }

    @JsonProperty("remote_data")
    public Optional<List<RemoteData>> getRemoteData() {
        return remoteData;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof PayrollRun && equalTo((PayrollRun) other);
    }

    private boolean equalTo(PayrollRun other) {
        return id.equals(other.id)
                && remoteId.equals(other.remoteId)
                && runState.equals(other.runState)
                && runType.equals(other.runType)
                && startDate.equals(other.startDate)
                && endDate.equals(other.endDate)
                && checkDate.equals(other.checkDate)
                && remoteWasDeleted.equals(other.remoteWasDeleted)
                && createdAt.equals(other.createdAt)
                && modifiedAt.equals(other.modifiedAt)
                && fieldMappings.equals(other.fieldMappings)
                && remoteData.equals(other.remoteData);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.id,
                this.remoteId,
                this.runState,
                this.runType,
                this.startDate,
                this.endDate,
                this.checkDate,
                this.remoteWasDeleted,
                this.createdAt,
                this.modifiedAt,
                this.fieldMappings,
                this.remoteData);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static Builder builder() {
        return new Builder();
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder {
        private Optional<String> id = Optional.empty();

        private Optional<String> remoteId = Optional.empty();

        private Optional<PayrollRunRunState> runState = Optional.empty();

        private Optional<PayrollRunRunType> runType = Optional.empty();

        private Optional<OffsetDateTime> startDate = Optional.empty();

        private Optional<OffsetDateTime> endDate = Optional.empty();

        private Optional<OffsetDateTime> checkDate = Optional.empty();

        private Optional<Boolean> remoteWasDeleted = Optional.empty();

        private Optional<OffsetDateTime> createdAt = Optional.empty();

        private Optional<OffsetDateTime> modifiedAt = Optional.empty();

        private Optional<Map<String, JsonNode>> fieldMappings = Optional.empty();

        private Optional<List<RemoteData>> remoteData = Optional.empty();

        private Builder() {}

        public Builder from(PayrollRun other) {
            id(other.getId());
            remoteId(other.getRemoteId());
            runState(other.getRunState());
            runType(other.getRunType());
            startDate(other.getStartDate());
            endDate(other.getEndDate());
            checkDate(other.getCheckDate());
            remoteWasDeleted(other.getRemoteWasDeleted());
            createdAt(other.getCreatedAt());
            modifiedAt(other.getModifiedAt());
            fieldMappings(other.getFieldMappings());
            remoteData(other.getRemoteData());
            return this;
        }

        @JsonSetter(value = "id", nulls = Nulls.SKIP)
        public Builder id(Optional<String> id) {
            this.id = id;
            return this;
        }

        public Builder id(String id) {
            this.id = Optional.of(id);
            return this;
        }

        @JsonSetter(value = "remote_id", nulls = Nulls.SKIP)
        public Builder remoteId(Optional<String> remoteId) {
            this.remoteId = remoteId;
            return this;
        }

        public Builder remoteId(String remoteId) {
            this.remoteId = Optional.of(remoteId);
            return this;
        }

        @JsonSetter(value = "run_state", nulls = Nulls.SKIP)
        public Builder runState(Optional<PayrollRunRunState> runState) {
            this.runState = runState;
            return this;
        }

        public Builder runState(PayrollRunRunState runState) {
            this.runState = Optional.of(runState);
            return this;
        }

        @JsonSetter(value = "run_type", nulls = Nulls.SKIP)
        public Builder runType(Optional<PayrollRunRunType> runType) {
            this.runType = runType;
            return this;
        }

        public Builder runType(PayrollRunRunType runType) {
            this.runType = Optional.of(runType);
            return this;
        }

        @JsonSetter(value = "start_date", nulls = Nulls.SKIP)
        public Builder startDate(Optional<OffsetDateTime> startDate) {
            this.startDate = startDate;
            return this;
        }

        public Builder startDate(OffsetDateTime startDate) {
            this.startDate = Optional.of(startDate);
            return this;
        }

        @JsonSetter(value = "end_date", nulls = Nulls.SKIP)
        public Builder endDate(Optional<OffsetDateTime> endDate) {
            this.endDate = endDate;
            return this;
        }

        public Builder endDate(OffsetDateTime endDate) {
            this.endDate = Optional.of(endDate);
            return this;
        }

        @JsonSetter(value = "check_date", nulls = Nulls.SKIP)
        public Builder checkDate(Optional<OffsetDateTime> checkDate) {
            this.checkDate = checkDate;
            return this;
        }

        public Builder checkDate(OffsetDateTime checkDate) {
            this.checkDate = Optional.of(checkDate);
            return this;
        }

        @JsonSetter(value = "remote_was_deleted", nulls = Nulls.SKIP)
        public Builder remoteWasDeleted(Optional<Boolean> remoteWasDeleted) {
            this.remoteWasDeleted = remoteWasDeleted;
            return this;
        }

        public Builder remoteWasDeleted(Boolean remoteWasDeleted) {
            this.remoteWasDeleted = Optional.of(remoteWasDeleted);
            return this;
        }

        @JsonSetter(value = "created_at", nulls = Nulls.SKIP)
        public Builder createdAt(Optional<OffsetDateTime> createdAt) {
            this.createdAt = createdAt;
            return this;
        }

        public Builder createdAt(OffsetDateTime createdAt) {
            this.createdAt = Optional.of(createdAt);
            return this;
        }

        @JsonSetter(value = "modified_at", nulls = Nulls.SKIP)
        public Builder modifiedAt(Optional<OffsetDateTime> modifiedAt) {
            this.modifiedAt = modifiedAt;
            return this;
        }

        public Builder modifiedAt(OffsetDateTime modifiedAt) {
            this.modifiedAt = Optional.of(modifiedAt);
            return this;
        }

        @JsonSetter(value = "field_mappings", nulls = Nulls.SKIP)
        public Builder fieldMappings(Optional<Map<String, JsonNode>> fieldMappings) {
            this.fieldMappings = fieldMappings;
            return this;
        }

        public Builder fieldMappings(Map<String, JsonNode> fieldMappings) {
            this.fieldMappings = Optional.of(fieldMappings);
            return this;
        }

        @JsonSetter(value = "remote_data", nulls = Nulls.SKIP)
        public Builder remoteData(Optional<List<RemoteData>> remoteData) {
            this.remoteData = remoteData;
            return this;
        }

        public Builder remoteData(List<RemoteData> remoteData) {
            this.remoteData = Optional.of(remoteData);
            return this;
        }

        public PayrollRun build() {
            return new PayrollRun(
                    id,
                    remoteId,
                    runState,
                    runType,
                    startDate,
                    endDate,
                    checkDate,
                    remoteWasDeleted,
                    createdAt,
                    modifiedAt,
                    fieldMappings,
                    remoteData);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/hris/PayrollRun.kt
package dev.merge.api.models.hris

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.errors.MergeInvalidDataException
import dev.merge.api.models.*
import java.time.OffsetDateTime
import java.util.Objects
import java.util.Optional

/**
 * # The PayrollRun Object
 *
 * ### Description
 *
 * The `PayrollRun` object is used to represent a group of pay statements for a specific pay
 * schedule.
 *
 * ### Usage Example
 *
 * Fetch from the `LIST PayrollRuns` endpoint and filter by `ID` to show all payroll runs.
 */
@JsonDeserialize(builder = PayrollRun.Builder::class)
@NoAutoDetect
class PayrollRun
private constructor(
    private val id: JsonField<String>,
    private val remoteId: JsonField<String>,
    private val runState: JsonField<RunStateEnum>,
    private val runType: JsonField<RunTypeEnum>,
    private val startDate: JsonField<OffsetDateTime>,
    private val endDate: JsonField<OffsetDateTime>,
    private val checkDate: JsonField<OffsetDateTime>,
    private val remoteWasDeleted: JsonField<Boolean>,
    private val fieldMappings: JsonValue,
    private val modifiedAt: JsonField<OffsetDateTime>,
    private val remoteData: JsonField<List<RemoteData>>,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun id(): Optional<String> = Optional.ofNullable(id.getNullable("id"))

    /** The third-party API ID of the matching object. */
    fun remoteId(): Optional<String> = Optional.ofNullable(remoteId.getNullable("remote_id"))

    /**
     * - `PAID` - PAID
     * - `DRAFT` - DRAFT
     * - `APPROVED` - APPROVED
     * - `FAILED` - FAILED
     * - `CLOSED` - CLOSED
     */
    fun runState(): Optional<RunStateEnum> = Optional.ofNullable(runState.getNullable("run_state"))

    /**
     * - `REGULAR` - REGULAR
     * - `OFF_CYCLE` - OFF_CYCLE
     * - `CORRECTION` - CORRECTION
     * - `TERMINATION` - TERMINATION
     * - `SIGN_ON_BONUS` - SIGN_ON_BONUS
     */
    fun runType(): Optional<RunTypeEnum> = Optional.ofNullable(runType.getNullable("run_type"))

    /** The day and time the payroll run started. */
    fun startDate(): Optional<OffsetDateTime> =
        Optional.ofNullable(startDate.getNullable("start_date"))

    /** The day and time the payroll run ended. */
    fun endDate(): Optional<OffsetDateTime> = Optional.ofNullable(endDate.getNullable("end_date"))

    /** The day and time the payroll run was checked. */
    fun checkDate(): Optional<OffsetDateTime> =
        Optional.ofNullable(checkDate.getNullable("check_date"))

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    fun remoteWasDeleted(): Optional<Boolean> =
        Optional.ofNullable(remoteWasDeleted.getNullable("remote_was_deleted"))

    /** This is the datetime that this object was last updated by Merge */
    fun modifiedAt(): Optional<OffsetDateTime> =
        Optional.ofNullable(modifiedAt.getNullable("modified_at"))

    fun remoteData(): Optional<List<RemoteData>> =
        Optional.ofNullable(remoteData.getNullable("remote_data"))

    @JsonProperty("id") @ExcludeMissing fun _id() = id

    /** The third-party API ID of the matching object. */
    @JsonProperty("remote_id") @ExcludeMissing fun _remoteId() = remoteId

    /**
     * - `PAID` - PAID
     * - `DRAFT` - DRAFT
     * - `APPROVED` - APPROVED
     * - `FAILED` - FAILED
     * - `CLOSED` - CLOSED
     */
    @JsonProperty("run_state") @ExcludeMissing fun _runState() = runState

    /**
     * - `REGULAR` - REGULAR
     * - `OFF_CYCLE` - OFF_CYCLE
     * - `CORRECTION` - CORRECTION
     * - `TERMINATION` - TERMINATION
     * - `SIGN_ON_BONUS` - SIGN_ON_BONUS
     */
    @JsonProperty("run_type") @ExcludeMissing fun _runType() = runType

    /** The day and time the payroll run started. */
    @JsonProperty("start_date") @ExcludeMissing fun _startDate() = startDate

    /** The day and time the payroll run ended. */
    @JsonProperty("end_date") @ExcludeMissing fun _endDate() = endDate

    /** The day and time the payroll run was checked. */
    @JsonProperty("check_date") @ExcludeMissing fun _checkDate() = checkDate

    /** Indicates whether or not this object has been deleted by third party webhooks. */
    @JsonProperty("remote_was_deleted") @ExcludeMissing fun _remoteWasDeleted() = remoteWasDeleted

    @JsonProperty("field_mappings") @ExcludeMissing fun _fieldMappings() = fieldMappings

    /** This is the datetime that this object was last updated by Merge */
    @JsonProperty("modified_at") @ExcludeMissing fun _modifiedAt() = modifiedAt

    @JsonProperty("remote_data") @ExcludeMissing fun _remoteData() = remoteData

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate(): PayrollRun = apply {
        if (!validated) {
            id()
            remoteId()
            runState()
            runType()
            startDate()
            endDate()
            checkDate()
            remoteWasDeleted()
            modifiedAt()
            remoteData().map { it.forEach { it.validate() } }
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is PayrollRun &&
            this.id == other.id &&
            this.remoteId == other.remoteId &&
            this.runState == other.runState &&
            this.runType == other.runType &&
            this.startDate == other.startDate &&
            this.endDate == other.endDate &&
            this.checkDate == other.checkDate &&
            this.remoteWasDeleted == other.remoteWasDeleted &&
            this.fieldMappings == other.fieldMappings &&
            this.modifiedAt == other.modifiedAt &&
            this.remoteData == other.remoteData &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    id,
                    remoteId,
                    runState,
                    runType,
                    startDate,
                    endDate,
                    checkDate,
                    remoteWasDeleted,
                    fieldMappings,
                    modifiedAt,
                    remoteData,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "PayrollRun{id=$id, remoteId=$remoteId, runState=$runState, runType=$runType, startDate=$startDate, endDate=$endDate, checkDate=$checkDate, remoteWasDeleted=$remoteWasDeleted, fieldMappings=$fieldMappings, modifiedAt=$modifiedAt, remoteData=$remoteData, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var id: JsonField<String> = JsonMissing.of()
        private var remoteId: JsonField<String> = JsonMissing.of()
        private var runState: JsonField<RunStateEnum> = JsonMissing.of()
        private var runType: JsonField<RunTypeEnum> = JsonMissing.of()
        private var startDate: JsonField<OffsetDateTime> = JsonMissing.of()
        private var endDate: JsonField<OffsetDateTime> = JsonMissing.of()
        private var checkDate: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteWasDeleted: JsonField<Boolean> = JsonMissing.of()
        private var fieldMappings: JsonValue = JsonMissing.of()
        private var modifiedAt: JsonField<OffsetDateTime> = JsonMissing.of()
        private var remoteData: JsonField<List<RemoteData>> = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(payrollRun: PayrollRun) = apply {
            this.id = payrollRun.id
            this.remoteId = payrollRun.remoteId
            this.runState = payrollRun.runState
            this.runType = payrollRun.runType
            this.startDate = payrollRun.startDate
            this.endDate = payrollRun.endDate
            this.checkDate = payrollRun.checkDate
            this.remoteWasDeleted = payrollRun.remoteWasDeleted
            this.fieldMappings = payrollRun.fieldMappings
            this.modifiedAt = payrollRun.modifiedAt
            this.remoteData = payrollRun.remoteData
            additionalProperties(payrollRun.additionalProperties)
        }

        fun id(id: String) = id(JsonField.of(id))

        @JsonProperty("id") @ExcludeMissing fun id(id: JsonField<String>) = apply { this.id = id }

        /** The third-party API ID of the matching object. */
        fun remoteId(remoteId: String) = remoteId(JsonField.of(remoteId))

        /** The third-party API ID of the matching object. */
        @JsonProperty("remote_id")
        @ExcludeMissing
        fun remoteId(remoteId: JsonField<String>) = apply { this.remoteId = remoteId }

        /**
         * - `PAID` - PAID
         * - `DRAFT` - DRAFT
         * - `APPROVED` - APPROVED
         * - `FAILED` - FAILED
         * - `CLOSED` - CLOSED
         */
        fun runState(runState: RunStateEnum) = runState(JsonField.of(runState))

        /**
         * - `PAID` - PAID
         * - `DRAFT` - DRAFT
         * - `APPROVED` - APPROVED
         * - `FAILED` - FAILED
         * - `CLOSED` - CLOSED
         */
        @JsonProperty("run_state")
        @ExcludeMissing
        fun runState(runState: JsonField<RunStateEnum>) = apply { this.runState = runState }

        /**
         * - `REGULAR` - REGULAR
         * - `OFF_CYCLE` - OFF_CYCLE
         * - `CORRECTION` - CORRECTION
         * - `TERMINATION` - TERMINATION
         * - `SIGN_ON_BONUS` - SIGN_ON_BONUS
         */
        fun runType(runType: RunTypeEnum) = runType(JsonField.of(runType))

        /**
         * - `REGULAR` - REGULAR
         * - `OFF_CYCLE` - OFF_CYCLE
         * - `CORRECTION` - CORRECTION
         * - `TERMINATION` - TERMINATION
         * - `SIGN_ON_BONUS` - SIGN_ON_BONUS
         */
        @JsonProperty("run_type")
        @ExcludeMissing
        fun runType(runType: JsonField<RunTypeEnum>) = apply { this.runType = runType }

        /** The day and time the payroll run started. */
        fun startDate(startDate: OffsetDateTime) = startDate(JsonField.of(startDate))

        /** The day and time the payroll run started. */
        @JsonProperty("start_date")
        @ExcludeMissing
        fun startDate(startDate: JsonField<OffsetDateTime>) = apply { this.startDate = startDate }

        /** The day and time the payroll run ended. */
        fun endDate(endDate: OffsetDateTime) = endDate(JsonField.of(endDate))

        /** The day and time the payroll run ended. */
        @JsonProperty("end_date")
        @ExcludeMissing
        fun endDate(endDate: JsonField<OffsetDateTime>) = apply { this.endDate = endDate }

        /** The day and time the payroll run was checked. */
        fun checkDate(checkDate: OffsetDateTime) = checkDate(JsonField.of(checkDate))

        /** The day and time the payroll run was checked. */
        @JsonProperty("check_date")
        @ExcludeMissing
        fun checkDate(checkDate: JsonField<OffsetDateTime>) = apply { this.checkDate = checkDate }

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        fun remoteWasDeleted(remoteWasDeleted: Boolean) =
            remoteWasDeleted(JsonField.of(remoteWasDeleted))

        /** Indicates whether or not this object has been deleted by third party webhooks. */
        @JsonProperty("remote_was_deleted")
        @ExcludeMissing
        fun remoteWasDeleted(remoteWasDeleted: JsonField<Boolean>) = apply {
            this.remoteWasDeleted = remoteWasDeleted
        }

        @JsonProperty("field_mappings")
        @ExcludeMissing
        fun fieldMappings(fieldMappings: JsonValue) = apply { this.fieldMappings = fieldMappings }

        /** This is the datetime that this object was last updated by Merge */
        fun modifiedAt(modifiedAt: OffsetDateTime) = modifiedAt(JsonField.of(modifiedAt))

        /** This is the datetime that this object was last updated by Merge */
        @JsonProperty("modified_at")
        @ExcludeMissing
        fun modifiedAt(modifiedAt: JsonField<OffsetDateTime>) = apply {
            this.modifiedAt = modifiedAt
        }

        fun remoteData(remoteData: List<RemoteData>) = remoteData(JsonField.of(remoteData))

        @JsonProperty("remote_data")
        @ExcludeMissing
        fun remoteData(remoteData: JsonField<List<RemoteData>>) = apply {
            this.remoteData = remoteData
        }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): PayrollRun =
            PayrollRun(
                id,
                remoteId,
                runState,
                runType,
                startDate,
                endDate,
                checkDate,
                remoteWasDeleted,
                fieldMappings,
                modifiedAt,
                remoteData.map { it.toUnmodifiable() },
                additionalProperties.toUnmodifiable(),
            )
    }

    class RunStateEnum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is RunStateEnum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val PAID = RunStateEnum(JsonField.of("PAID"))

            @JvmField val DRAFT = RunStateEnum(JsonField.of("DRAFT"))

            @JvmField val APPROVED = RunStateEnum(JsonField.of("APPROVED"))

            @JvmField val FAILED = RunStateEnum(JsonField.of("FAILED"))

            @JvmField val CLOSED = RunStateEnum(JsonField.of("CLOSED"))

            @JvmStatic fun of(value: String) = RunStateEnum(JsonField.of(value))
        }

        enum class Known {
            PAID,
            DRAFT,
            APPROVED,
            FAILED,
            CLOSED,
        }

        enum class Value {
            PAID,
            DRAFT,
            APPROVED,
            FAILED,
            CLOSED,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                PAID -> Value.PAID
                DRAFT -> Value.DRAFT
                APPROVED -> Value.APPROVED
                FAILED -> Value.FAILED
                CLOSED -> Value.CLOSED
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                PAID -> Known.PAID
                DRAFT -> Known.DRAFT
                APPROVED -> Known.APPROVED
                FAILED -> Known.FAILED
                CLOSED -> Known.CLOSED
                else -> throw MergeInvalidDataException("Unknown RunStateEnum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }

    class RunTypeEnum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is RunTypeEnum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val REGULAR = RunTypeEnum(JsonField.of("REGULAR"))

            @JvmField val OFF_CYCLE = RunTypeEnum(JsonField.of("OFF_CYCLE"))

            @JvmField val CORRECTION = RunTypeEnum(JsonField.of("CORRECTION"))

            @JvmField val TERMINATION = RunTypeEnum(JsonField.of("TERMINATION"))

            @JvmField val SIGN_ON_BONUS = RunTypeEnum(JsonField.of("SIGN_ON_BONUS"))

            @JvmStatic fun of(value: String) = RunTypeEnum(JsonField.of(value))
        }

        enum class Known {
            REGULAR,
            OFF_CYCLE,
            CORRECTION,
            TERMINATION,
            SIGN_ON_BONUS,
        }

        enum class Value {
            REGULAR,
            OFF_CYCLE,
            CORRECTION,
            TERMINATION,
            SIGN_ON_BONUS,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                REGULAR -> Value.REGULAR
                OFF_CYCLE -> Value.OFF_CYCLE
                CORRECTION -> Value.CORRECTION
                TERMINATION -> Value.TERMINATION
                SIGN_ON_BONUS -> Value.SIGN_ON_BONUS
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                REGULAR -> Known.REGULAR
                OFF_CYCLE -> Known.OFF_CYCLE
                CORRECTION -> Known.CORRECTION
                TERMINATION -> Known.TERMINATION
                SIGN_ON_BONUS -> Known.SIGN_ON_BONUS
                else -> throw MergeInvalidDataException("Unknown RunTypeEnum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }
}
